v2/yarpcthrift/codec_test.go

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package yarpcthrift
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/thriftrw/protocol"
+	"go.uber.org/thriftrw/thrifttest"
+	"go.uber.org/thriftrw/wire"
+	yarpc "go.uber.org/yarpc/v2"
+)
+
+func TestDecodeAgnosticProto(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	proto := thrifttest.NewMockEnvelopeAgnosticProtocol(mockCtrl)
+	proto.EXPECT().DecodeRequest(wire.Call, gomock.Any()).Return(
+		wire.NewValueStruct(wire.Struct{}), protocol.NoEnvelopeResponder, nil).Times(2)
+
+	testCodec := newCodec(proto, false)
+	reqBuf := &yarpc.Buffer{}
+
+	res, err := testCodec.Decode(reqBuf)
+	assert.NoError(t, err, "unexpected error")
+	resVal, ok := res.(wire.Value)
+	assert.True(t, ok, "expected wire.Value")
+	assert.Equal(t, wire.NewValueStruct(wire.Struct{}), resVal)
+
+	// calling Decode again should fail, because it attempts to set a new
+	// responder.
+	res, err = testCodec.Decode(reqBuf)
+	assert.Error(t, err, "unexpected error")
+}
+
+func TestDecodeAgnosticProtoError(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	proto := thrifttest.NewMockEnvelopeAgnosticProtocol(mockCtrl)
+	proto.EXPECT().DecodeRequest(wire.Call, gomock.Any()).Return(
+		wire.Value{}, nil, errors.New("error decoding request"))
+
+	testCodec := newCodec(proto, false)
+	reqBuf := &yarpc.Buffer{}
+
+	res, err := testCodec.Decode(reqBuf)
+	assert.Error(t, err, "expected error")
+	assert.Nil(t, res)
+}
```

Comment:
optional: Many of these tests take on the same shape. In these cases, it might make more sense to use table-driven tests. 

Not necessarily actionable here, since I realize that these tests were largely referenced from the previous implementation.